### ISSUE TITLE
How to deprecate a library?

### ISSUE THREAD

**Issue by maximedenes** [coq/coq#8032]
In order to provide smooth transition paths to our users, I'd like to implement a deprecation mechanism for libraries. That is, the possibility to emit a warning when a user `Require`s a given `.vo` file (directly, not through indirect dependencies).

Any suggestion on the proper way to do that?

**Comment by ppedrot** [comment]
Without consideration for other features that could come bundled with this one, a natural way for me would be to provide an additional flag to `coqc` at compilation time that would store this information in the object file.

**Comment by maximedenes** [comment]
> Let me rephrase, is a library a module too?

AFAIK, some parts of the system consider it as a module, some other parts consider it as a special construct...

But note that my question does not make sense on modules, since they can't be `Require`d.

Why do you want an answer to that question? I.e. what would "yes, it is a module" or "no, it is no" imply on the deprecation mechanism, in your view?

**Comment by ejgallego** [comment]
As you mentioned Haskell I feel that first we ought to think a bit about the strange model we have with `Require` + ` Import`. Neither Haskell nor OCaml have these two constructs, and indeed the whole `Require` business has always seemed suspicious to me. Certainly adding yet another keyword `Library` is not a favorite option. A top-level attribute seems like a better choice as it would cover more use cases.

As mentioned in Gitter, transitivity may be an issue.

I also wonder about the use case itself; what is the point of deprecating a whole library? That seems quite coarse-grained to me; if the replacement is compatible, you could just remap, and if it is not, why does the old lib need to go away; you could just put it in a differente package and live with it, wouldn't you?

**Comment by maximedenes** [comment]
> I also wonder about the use case itself; what is the point of deprecating a whole library? That seems quite coarse-grained to me; if the replacement is compatible, you could just remap, and if it is not, why does the old lib need to go away; you could just put it in a differente package and live with it, wouldn't you?

What do you concretely mean by "remapping"? The point of the warning is precisely to be able to remove the remapping at some point.

**Comment by maximedenes** [comment]
> As you mentioned Haskell I feel that first we ought to think a bit about the strange model we have with Require + Import. Neither Haskell nor OCaml have these two constructs, and indeed the whole Require business has always seemed suspicious to me.

Indeed, the model is a bit uncommon. In many languages, `import` supports a fine-grained selection of what is to be imported, and declares the dependency at the same time (a bit like `Require` does today).

@ppedrot do you know why we have `Require`?

**Comment by ejgallego** [comment]
Well we have require as to be able to perform side-effects without opening a namespace. But as we all know this is a bit messy as of today; in particular in terms of implementation.

**Comment by ejgallego** [comment]
> The point of the warning is precisely to be able to remove the remapping at some point.

Yeah so maybe a top-level attribute would work well for this; on the other hand using packages could work better, I guess we'd need to see a concrete use case to know better.

**Comment by maximedenes** [comment]
> I guess we'd need to see a concrete use case to know better

Take for example `Fourier.vo` in #7897.

**Comment by ejgallego** [comment]
> Take for example Fourier.vo in #7897.

A top-level attribute would work well; problem is that `.v` files are not so consistently parsed from a top-level file specification. In a sense this is what you are asking about having a top-level "Library" stanza; I would just reuse module and take the opportunity to clarify what side-effects do modules can carry.

This made me think that we should indeed start to structure `Require` side-effects on `.v` files.

**Comment by ejgallego** [comment]
The thing is that doing this properly is going to require significant design effort.

**Comment by maximedenes** [comment]
> needing to perform the side effects

This is long term indeed, but sometimes good to think about it. What side effects do we really want to perform at `Require` time, in an ideal system?

> I guess that for Fourier you could just add something such as:
> 
> #[deprecated] Require Export FourierSubstitute.

When would the warning be printed?

**Comment by ejgallego** [comment]
> When would the warning be printed?

That should print it at `.vo` Require time I think; the code that avoids double loading should prevent the warning to be printed twice, which is OK but not ideal in some scenarios.



> What side effects do we really want to perform at Require time, in an ideal system?

The thing is that it varies I think! Each users will want a different combination.

Maybe indeed we should get rid of `Require` completely and instead allow to do `Import {notations, hints} From Foo.` etc...

**Comment by maximedenes** [comment]
> Maybe indeed we should get rid of Require completely and instead allow to do Import {notations, hints} From Foo. etc...

Yeah, this is what I was wondering.

> That should print it at .vo Require time I think;

Trying to make sure I understand: the deprecation annotation on `Require Export` would trigger a deprecation warning when `Requiring` the file that contains it?

**Comment by ejgallego** [comment]
> Trying to make sure I understand: the deprecation annotation on Require Export would trigger a deprecation warning when Requiring the file that contains it?

This is just a conjecture, but given that `Require Export` is a side effect on itself, it should be triggered on `Require` I think.

**Comment by maximedenes** [comment]
> In a sense this is what you are asking about having a top-level "Library" stanza; I would just reuse module and take the opportunity to clarify what side-effects do modules can carry.

And by this you mean that the long term would be to use a mandatory toplevel `Module` that designates the file, à la Haskell, right?

My suggestion to use a separate vernacular was just because I don't see how the transition strategy would go otherwise.

**Comment by ejgallego** [comment]
> And by this you mean that the long term would be to use a mandatory toplevel Module that designates the file, à la Haskell, right?

That could be a possibility, tho if that stanza is missing it can be always inserted implicitly [and warning emitted if we desire tho].

But I think the crucial thing here is to organize the side-effects in modules / libraries; as it is now its scope is hard to control even in modules so repurposing `Import/Export` for that could make sense.

An added difficulty is that basically plugins can do what they want here; but some namespacing could be enforced. Anyways I am still not very optimistic with plugins, for these issues and also for some deeper problems such as #7698 .

**Comment by maximedenes** [comment]
> That could be a possibility, tho if that stanza is missing it can be always inserted implicitly [and warning emitted if we desire tho].

Sure, but what I don't see as far as transition is concerned, is what to do for files that already start with `Module` today, which will then change semantics. Of course, it is probably a rare occurrence, so we may be incompatible there.

> But I think the crucial thing here is to organize the side-effects in modules / libraries; as it is now its scope is hard to control even in modules so repurposing Import/Export for that could make sense.

Indeed, the clean-up work to be done for effects in modules is probably huge, but we will probably have to do it at some point.

**Comment by ejgallego** [comment]
> Sure, but what I don't see as far as transition is concerned, is what to do for files that already start with Module today, which will then change semantics. Of course, it is probably a rare occurrence, so we may be incompatible there.

Oh good point, I guess we could add `Library` or something else [maybe an attribute] but as long as it is not really a new command but just an alias for `Module`.

Another trick to do the deprecation right now could be to do:
```
#[deprecated "msg"] Comments "this file is deprecated, use Foo".
```

**Comment by maximedenes** [comment]
> #[deprecated "msg"] Comments "this file is deprecated, use Foo".

Ah ah, that's an interesting suggestion, thanks, I'll think about it.

**Comment by gares** [comment]
I think the `Comments` thing is the way to go (eg a "dummy" libobject item in the .vo file), but the two messages are confusing (eg "use Foo" should be in the attribute).

**Comment by ejgallego** [comment]
Indeed I have thought a bit more about this and I am not sure the proposals I've made are OK.

I'll think a bit more about it, and indeed, read again the libobject code.

**Comment by ejgallego** [comment]
cc #9811

**Comment by Zimmi48** [comment]
In https://github.com/coq/coq/pull/13741#issuecomment-764632075, @RalfJung mentioned that Rust's syntax for attributes that are attached to the current scope / module is `#![attribute]`.

**Comment by RalfJung** [comment]
Right, so in Rust, to deprecate the of file module `foo.rs`, one would have something like this at the top of the file
```
#![deprecated(since = "version", note = "xxx")]
```

**Comment by herbelin** [comment]
#18193 is an experiment to associate a warning (or deprecation) to a library.

Hopefully, it can reconcile the diversity of point of views expressed above:
- a comment can be attached to all declarations, and in particular to modules (in particular, a file is a `MPfile` module)
- the printing of the warning is attached to the `Require`
- the declaration of the warning is like an "attribute" in (possibly the beginning of) the file (but it could also be a coqc option)
